src/vsr/replica.zig

```diff
@@ -2263,25 +2261,27 @@ pub fn ReplicaType(
 
             assert(read.destination != self.replica);
 
-            result catch {
-                log.debug("{}: on_request_blocks: block not found " ++
+            if (result != .valid) {
+                log.debug("{}: on_request_blocks: error: {s}: " ++
                     "(destination={} address={} checksum={})", .{
                     self.replica,
+                    @tagName(result),
                     read.destination,
                     grid_read.address,
                     grid_read.checksum,
                 });
                 return;
-            };
+            }
 
-            log.debug("{}: on_request_blocks: block found " ++
-                "(destination={} address={} checksum={})", .{
+            log.debug("{}: on_request_blocks: success: (destination={} address={} checksum={})", .{
                 self.replica,
                 read.destination,
                 grid_read.address,
                 grid_read.checksum,
             });
 
+            stdx.copy_disjoint(.inexact, u8, read.message.buffer, result.valid);
```

Comment:
It feels like we should be able to eliminate this copy in theory, as we could read into this buffer in the first place. 

If we go for refcounted blocks, we could even make block=message

Reply:
Interesting! :thinking: 
So the `MessagePool` would have a bunch of `Message`s, some of which are 1MiB and some which are 64KiB.
Ironically this is similar to the very original messagepool design, which treated header-only messages specially.